src/mlpack/methods/bayesian_linear_regression/bayesian_linear_regression.cpp

```diff
@@ -0,0 +1,175 @@
+
+  if (!arma::eig_sym(eigVal, eigVec, arma::symmatu(phi * phi.t())))
+  {
+    Log::Fatal << "BayesianLinearRegression::Train(): Eigendecomposition "
+               << "of covariance failed!";
+  }
+
+  // Compute this quantities once and for all.
+  const arma::mat eigVecInv = inv(eigVec);
+  const arma::colvec eigVecInvPhitT = eigVecInv * phi * t.t();
+
+  // Initialize the hyperparameters and
+  // begin with an infinitely broad prior.
+  alpha = 1e-6;
+  beta =  1 / (var(t, 1) * 0.1);
+
+  unsigned short i = 0;
+  double deltaAlpha = 1.0, deltaBeta = 1.0, crit = 1.0;
+
+  while ((crit > tol) && (i < nIterMax))
+  {
+    deltaAlpha = -alpha;
+    deltaBeta = -beta;
+
+    // Update the solution.
+    omega = eigVec * diagmat(1 / (eigVal + (alpha / beta))) * eigVecInvPhitT;
+
+    // Update alpha.
+    gamma = sum(eigVal / (alpha / beta + eigVal));
+    alpha = gamma / dot(omega, omega);
+
+    // Update beta.
+    const arma::rowvec temp = t - omega.t() * phi;
+    beta = (data.n_cols - gamma) / dot(temp, temp);
+
+    // Compute the stopping criterion.
+    deltaAlpha += alpha;
+    deltaBeta += beta;
+    crit = std::abs(deltaAlpha / alpha + deltaBeta / beta);
+    i++;
+  }
+  // Compute the covariance matrix for the uncertainties later.
+  matCovariance = eigVec * diagmat(1 / (beta * eigVal + alpha)) * eigVecInv;
+
+  Timer::Stop("bayesian_linear_regression");
+
+  return RMSE(data, responses);
+}
+
+void BayesianLinearRegression::Predict(const arma::mat& points,
+                                       arma::rowvec& predictions) const
+{
+  // Center and scale the points before applying the model.
+  arma::mat matX;
+  CenterScaleDataPred(points, matX);
+  predictions = omega.t() * matX + responsesOffset;
+}
+
+void BayesianLinearRegression::Predict(const arma::mat& points,
+                                       arma::rowvec& predictions,
+                                       arma::rowvec& std) const
+{
+  // Center and scale the points before applying the model.
+  arma::mat matX;
+  CenterScaleDataPred(points, matX);
+  predictions = omega.t() * matX + responsesOffset;
+  // Compute the standard deviation dor each points.
+  std = sqrt(Variance() + sum((matX % (matCovariance * matX)), 0));
+}
+
+double BayesianLinearRegression::RMSE(const arma::mat& data,
+                                      const arma::rowvec& responses) const
+{
+  arma::rowvec predictions;
+  Predict(data, predictions);
+  return sqrt(mean(square(responses - predictions)));
+}
+
+double BayesianLinearRegression::CenterScaleData(const arma::mat& data,
+                                                 const arma::rowvec& responses,
+                                                 arma::mat& dataProc,
+                                                 arma::rowvec& responsesProc)
+{
+  // Initialize the offsets to their neutral forms.
+  responsesOffset = 0.0;
+  if (!centerData && !scaleData)
+  {
+    dataProc = data;
+    responsesProc = responses;
```

Comment:
The code you currently have will actually copy the data, but we could make it more efficient via the use of Armadillo aliases, which will make `dataProc` and `responsesProc` hold the same data as `data` and `responses` but without a copy. :+1:
```suggestion
    dataProc = arma::mat(const_cast<double*>(data.memptr()), data.n_rows, data.n_cols, false, true);
    responsesProc = arma::rowvec(const_cast<double*>(responses.memptr()), responses.n_elem, false, true);
```

Reply:
Very good +1: